staging/src/k8s.io/apiserver/plugin/pkg/audit/webhook/webhook.go

```diff
@@ -0,0 +1,285 @@
+	ModeBlocking = "blocking"
+)
+
+// AllowedModes is the modes known by this webhook.
+var AllowedModes = []string{
+	ModeBatch,
+	ModeBlocking,
+}
+
+const (
+	// Default configuration values for ModeBatch.
+	//
+	// TODO(ericchiang): Make these value configurable. Maybe through a
+	// kubeconfig extension?
+	defaultBatchBufferSize = 1000        // Buffer up to 1000 events before blocking.
+	defaultBatchMaxSize    = 100         // Only send 100 events at a time.
+	defaultBatchMaxWait    = time.Minute // Send events at least once a minute.
+)
+
+// NewBackend returns an audit backend that sends events over HTTP to an external service.
+// The mode indicates the caching behavior of the webhook. Either blocking (ModeBlocking)
+// or buffered with batch POSTs (ModeBatch).
+func NewBackend(kubeConfigFile string, mode string) (audit.Backend, error) {
+	switch mode {
+	case ModeBatch:
+		return newBatchWebhook(kubeConfigFile)
+	case ModeBlocking:
+		return newBlockingWebhook(kubeConfigFile)
+	default:
+		return nil, fmt.Errorf("webhook mode %q is not in list of known modes (%s)",
+			mode, strings.Join(AllowedModes, ","))
+	}
+}
+
+var (
+	// NOTE: Copied from other webhook implementations
+	//
+	// Can we make these passable to NewGenericWebhook?
+	groupFactoryRegistry = make(announced.APIGroupFactoryRegistry)
+	groupVersions        = []schema.GroupVersion{auditv1alpha1.SchemeGroupVersion}
+	registry             = registered.NewOrDie("")
+)
+
+func init() {
+	registry.RegisterVersions(groupVersions)
+	if err := registry.EnableVersions(groupVersions...); err != nil {
+		panic(fmt.Sprintf("failed to enable version %v", groupVersions))
+	}
+	install.Install(groupFactoryRegistry, registry, audit.Scheme)
+}
+
+func loadWebhook(configFile string) (*webhook.GenericWebhook, error) {
+	return webhook.NewGenericWebhook(registry, audit.Codecs, configFile, groupVersions, 0)
+}
+
+func newBlockingWebhook(configFile string) (*blockingBackend, error) {
+	w, err := loadWebhook(configFile)
+	if err != nil {
+		return nil, err
+	}
+	return &blockingBackend{w}, nil
+}
+
+type blockingBackend struct {
+	w *webhook.GenericWebhook
+}
+
+func (b *blockingBackend) Run(stopCh <-chan struct{}) error {
+	return nil
+}
+
+func (b *blockingBackend) ProcessEvents(ev ...*auditinternal.Event) {
+	if err := b.processEvents(ev...); err != nil {
+		glog.Errorf("failed to POST webhook events: %v", err)
+	}
+}
+
+func (b *blockingBackend) processEvents(ev ...*auditinternal.Event) error {
+	var list auditinternal.EventList
+	for _, e := range ev {
+		list.Items = append(list.Items, *e)
+	}
+	// NOTE: No exponential backoff because this is the blocking webhook
+	// mode. Any attempts to retry will block API server requests.
+	return b.w.RestClient.Post().Body(&list).Do().Error()
+}
+
+// Copied from generated code in k8s.io/apiserver/pkg/apis/audit.
+//
+// TODO(ericchiang): Have the generated code expose these methods like metav1.GetGeneratedDeepCopyFuncs().
```

Comment:
Going to follow up with a change to not have this copied from the audit package. It's currently in a generated file that I don't want to mess with:

https://github.com/kubernetes/apiserver/blob/c809cf8581e1e44c6174bf5ab4415e6ee39965ca/pkg/apis/audit/zz_generated.deepcopy.go#L37-L45

@sttts